components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListenerTest.java

```diff
@@ -0,0 +1,338 @@
+        serviceProvider.setApplicationID(1);
+        serviceProvider.setSaasApp(true);
+        when(mockAppMgtService.getServiceProvider(serviceProvider.getApplicationID()))
+                .thenReturn(persistedServiceProvider);
+
+        Boolean result = oAuthApplicationMgtListener.doPreUpdateApplication(serviceProvider, tenantDomain, userName);
+        assertTrue(result, "Pre-update application failed.");
+    }
+
+    @Test(dataProvider = "GetSPConfigData")
+    public void testDoPostGetServiceProvider(Boolean hasAuthConfig, Boolean hasRequestConfig, String authType,
+                                             String propName) throws Exception {
+
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+
+            ServiceProvider serviceProvider =
+                    createServiceProvider(1, hasAuthConfig, hasRequestConfig, authType, propName);
+            Boolean result =
+                    oAuthApplicationMgtListener.doPostGetServiceProvider(serviceProvider, spName, tenantDomain);
+            assertTrue(result, "Post-get service provider failed.");
+        }
+    }
+
+    @Test
+    public void testDoPostGetServiceProviderWhenSPisNull() throws Exception {
+
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+
+            Boolean result = oAuthApplicationMgtListener.doPostGetServiceProvider(null, spName, tenantDomain);
+            assertTrue(result, "Post-get service provider failed.");
+        }
+    }
+
+    @Test
+    public void testDoPostGetServiceProviderByClientId() throws Exception {
+
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+
+            ServiceProvider serviceProvider = createServiceProvider(1, true, true, OAUTH2, OAUTH_CONSUMER_SECRET);
+            Boolean result = oAuthApplicationMgtListener.doPostGetServiceProviderByClientId(serviceProvider,
+                    "clientId", "clientType", tenantDomain);
+            assertTrue(result, "Post-get service provider by client ID failed.");
+        }
+    }
+
+    @Test
+    public void testDoPostCreateApplication() throws Exception {
+
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+
+            ServiceProvider serviceProvider = createServiceProvider(1, true, true, OAUTH2, OAUTH_CONSUMER_SECRET);
+            Boolean result = oAuthApplicationMgtListener.doPostCreateApplication(serviceProvider, tenantDomain, userName);
+            assertTrue(result, "Post-create application failed.");
+        }
+    }
+
+    @DataProvider(name = "GetPostUpdateApplicationData")
+    public Object[][] PostUpdateApplicationData() {
+
+        return new Object[][]{
```

Comment:
Having comments to relate the data provided with how it relates to the tested flow would make the test much readable and easy to debug/refactor/fix the unit tests. WDYT?

Reply:
updated Data provider with comments on data entries.